# Workshop 1
--------------------------
This is codes used for Sydney's Python Workshop 1, on June 8th 2020. If you have any questions, feel free to email me (bs3222@columbia.edu)

#### 1. How to use markdown

In Python notebook, you can change the code chunk as "markdown". At jupyter notebook, on the far right, you will see a dropdown menu. Default is "code". You can change that to "markdown" and write markdown text.

Markdown is a plain-text-formatting syntax, used often for readme files.
You can understand it as text formatting.

If you add a hashtag (#) it becomes a header. If you add more hashtags, you will have smaller sized header.

# Header
## Header 2
### Header 3

If you have two underscores (\__)between text, it makes the text bold.
You can do the same with asterisk(\*) sign. If you have two asterisk signs between text, it makes the text bold.

__Make it bold with two underscores__ 

**Make it bold with two asterisk signs**

If you have one underscore or one asterisk sign, it makes the text italic.

_Make it italic with one underscore_ 

*Make it italic with one asterisk sign*

You can mix the two as well, like below.

__*text*__

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# os.getcwd() means "i) from os package ii) use the function getcwd iii) execute ()"
#  This function gets "current working directory"
os.getcwd() 

'C:\\Users\\bolim\\Google Drive\\DA_with_python\\python_workshop'

In [3]:
# A hashtag in a code chunk means that it is a "text", not "code".
# if you want to change your directory, use os.chdir() with the path specified inside the parenthesis
# os.chdir('directory_path/path/path') 
# This will not run because the code line is commented. (understood as text by python)

In [4]:
# Read in the csv file "surveys.csv", call that as d
d = pd.read_csv('surveys.csv')

In [6]:
#specify how many rows to read, write inside the parenthesis. default is 5 first rows
d.head(10) 

record_id  month  day  year  plot_id species_id sex  hindfoot_length  \
0          1      7   16  1977        2         NL   M             32.0   
1          2      7   16  1977        3         NL   M             33.0   
2          3      7   16  1977        2         DM   F             37.0   
3          4      7   16  1977        7         DM   M             36.0   
4          5      7   16  1977        3         DM   M             35.0   
5          6      7   16  1977        1         PF   M             14.0   
6          7      7   16  1977        2         PE   F              NaN   
7          8      7   16  1977        1         DM   M             37.0   
8          9      7   16  1977        1         DM   F             34.0   
9         10      7   16  1977        6         PF   F             20.0   

   weight  
0     NaN  
1     NaN  
2     NaN  
3     NaN  
4     NaN  
5     NaN  
6     NaN  
7     NaN  
8     NaN  
9     NaN

In [7]:
d.tail() # shows the bottom 5 rows. default also 5.

record_id  month  day  year  plot_id species_id  sex  hindfoot_length  \
35544      35545     12   31  2002       15         AH  NaN              NaN   
35545      35546     12   31  2002       15         AH  NaN              NaN   
35546      35547     12   31  2002       10         RM    F             15.0   
35547      35548     12   31  2002        7         DO    M             36.0   
35548      35549     12   31  2002        5        NaN  NaN              NaN   

       weight  
35544     NaN  
35545     NaN  
35546    14.0  
35547    51.0  
35548     NaN

In [8]:
# What is the type of d
type(d)

pandas.core.frame.DataFrame

`d` is pandas dataframe. 

In [9]:
# Get information of d
d.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35549 entries, 0 to 35548
Data columns (total 9 columns):
record_id          35549 non-null int64
month              35549 non-null int64
day                35549 non-null int64
year               35549 non-null int64
plot_id            35549 non-null int64
species_id         34786 non-null object
sex                33038 non-null object
hindfoot_length    31438 non-null float64
weight             32283 non-null float64
dtypes: float64(2), int64(5), object(2)
memory usage: 2.4+ MB


* `'pandas.core.frame.DataFrame'` = This is pandas dataframe
* `RangeIndex: 35549 entries, 0 to 35548` = The range of index is from 0 to 35548. There are 35549 entries. This means that there are 35549 rows in this dataset.
* It tells data type (dtype) of each column, and how many non null values exist for each column
* Column `record_id` has 35549 non-null values, and the data type is integer 64.
* Column `species_id` has 34786 non-null values, and the data type is object (character). As you can see, the number of non-null values does not equal to total number of rows in this dataset. This means that there _are_ null values for `species_id`.
* `float64` means that it is a number, with decimals. `int64` has no decimals.

In [12]:
# string/object example
example_object = "this is a string" # Use quotation marks, you can also use '' instead of ""

In [13]:
# check type of example object
type(example_object)

str

In [16]:
# Without the parenthesis after info, it will give the head and tail of the dataframe d
d.info

<bound method DataFrame.info of        record_id  month  day  year  plot_id species_id  sex  hindfoot_length  \
0              1      7   16  1977        2         NL    M             32.0   
1              2      7   16  1977        3         NL    M             33.0   
2              3      7   16  1977        2         DM    F             37.0   
3              4      7   16  1977        7         DM    M             36.0   
4              5      7   16  1977        3         DM    M             35.0   
...          ...    ...  ...   ...      ...        ...  ...              ...   
35544      35545     12   31  2002       15         AH  NaN              NaN   
35545      35546     12   31  2002       15         AH  NaN              NaN   
35546      35547     12   31  2002       10         RM    F             15.0   
35547      35548     12   31  2002        7         DO    M             36.0   
35548      35549     12   31  2002        5        NaN  NaN              NaN   

       

In [17]:
# Show column names
d.columns # column names

Index(['record_id', 'month', 'day', 'year', 'plot_id', 'species_id', 'sex',
       'hindfoot_length', 'weight'],
      dtype='object')

If the dataframe has a lot of columns, this will not print all the columns.
If you want to print each, you can use a `for` loop, which means to apply what you specify for each element.

In [18]:
# for each element (I will call that element "col") in d.columns
# print it "print(col)"
for col in d.columns:
    print(col)

record_id
month
day
year
plot_id
species_id
sex
hindfoot_length
weight


In [19]:
# I can change the print statement and add more things like this!
for col in d.columns:
    print("The columns are: ", col)

The columns are:  record_id
The columns are:  month
The columns are:  day
The columns are:  year
The columns are:  plot_id
The columns are:  species_id
The columns are:  sex
The columns are:  hindfoot_length
The columns are:  weight


In [20]:
# To get unique values.
# From pandas, use unique function. Inside the parenthesis, specify the column.
pd.unique(d['species_id'])

array(['NL', 'DM', 'PF', 'PE', 'DS', 'PP', 'SH', 'OT', 'DO', 'OX', 'SS',
       'OL', 'RM', nan, 'SA', 'PM', 'AH', 'DX', 'AB', 'CB', 'CM', 'CQ',
       'RF', 'PC', 'PG', 'PH', 'PU', 'CV', 'UR', 'UP', 'ZL', 'UL', 'CS',
       'SC', 'BA', 'SF', 'RO', 'AS', 'SO', 'PI', 'ST', 'CU', 'SU', 'RX',
       'PB', 'PL', 'PX', 'CT', 'US'], dtype=object)

In [22]:
# You can use dataframe.column_name instead of dataframe['column_name']
pd.unique(d.sex)

array(['M', 'F', nan], dtype=object)

* Some other helpful keyboard shortcuts.

1) To get all functions in a package, type the package and press "tab"

2) To get documentation of a function, use `help`. (ex: `help(pd.unique)`). Scroll down to the bottom and there are examples.

3) To run your code chunk, use `shift + enter`

4) If you want to run all your code chunks, you can go to the top tabs, cell > run all

In [1]:
# example of help(pd.unique)
help(pd.unique)

Help on function unique in module pandas.core.algorithms:

unique(values)
    Hash table-based unique. Uniques are returned in order
    of appearance. This does NOT sort.
    
    Significantly faster than numpy.unique. Includes NA values.
    
    Parameters
    ----------
    values : 1d array-like
    
    Returns
    -------
    numpy.ndarray or ExtensionArray
    
        The return can be:
    
        * Index : when the input is an Index
        * Categorical : when the input is a Categorical dtype
        * ndarray : when the input is a Series/ndarray
    
        Return numpy.ndarray or ExtensionArray.
    
    See Also
    --------
    Index.unique
    Series.unique
    
    Examples
    --------
    >>> pd.unique(pd.Series([2, 1, 3, 3]))
    array([2, 1, 3])
    
    >>> pd.unique(pd.Series([2] + [1] * 5))
    array([2, 1])
    
    >>> pd.unique(pd.Series([pd.Timestamp('20160101'),
    ...                     pd.Timestamp('20160101')]))
    array(['2016-01-01T00:00:00.0000